### ISSUE TITLE
Context.defineModule on exists module with TDField not works

### ISSUE THREAD

**Issue by lublak** [HaxeFoundation/haxe#10374]
I can use `Context.defineModule` to create a extra class in a module to inject some stuff.
This works fine.
If i use TDField the compiler throws an error: `File "src/typing/typeloadModule.ml", line 402, characters 4-10: Assertion failed`

```haxe
Context.defineModule('sub.test.Test', [{
  pos: Context.currentPos(),
  pack: ['sub', 'test'],
  name: 'SomeFancyVarName',
  params: [],
  meta: [],
  kind: TDField(FVar(null, macro null), [AInline, AFinal]),
  fields: []
}]);
```

**Comment by RealyUniqueName** [comment]
I cannot reproduce this issue.
Please, provide a complete sample and your Haxe version.

**Comment by nadako** [comment]
I could reproduce this and it's a bit annoying to fix because it involves adding fields to an already existing class which might be already built...

**Comment by lublak** [comment]
@nadako  So from my point of view, it doesn't have to be mandatory. Now I simply declare a class with a field and return the path as an expression.
I only noticed it because I am currently cleaning up my entire code from scratch and bringing it to Haxe 4.
An error message that it is not possible instead of typeloader.ml would make sense.
If it is implemented, of course, it is also great. But maybe it's more of a special case.

**Comment by RealyUniqueName** [comment]
I still need a sample which I can use to reproduce it myself to be able to fix anything.

**Comment by lublak** [comment]
@RealyUniqueName I can try to make one tomorrow.
I try to make it as small as possible.

**Comment by lublak** [comment]
@RealyUniqueName here ist an example:

Main.hx
```haxe
import haxe.macro.Context;

class Main {
	static function main() {
		new Test();
		test();
	}

	static macro function test() {
		if(try {Context.getType('Test.Sub'); false;} catch(_:Any) true) {
			Context.defineModule('Test', [{
				pos: Context.currentPos(),
				pack: [],
				name: 'Sub',
				params: [],
				meta: [],
				kind: TDField(FVar(null, macro null), [AInline, AFinal]),
				fields: []
			}]);
		}
		
		return macro macro {};
	}
}
```

Test.hx
```haxe
#if !macro @:build(TestBuilder.build()) #end
class Test {
  public function new() {
    
  }
}
```

TestBuilder.hx
```haxe
import haxe.macro.Context;

class TestBuilder {
  public static function build() {
    return Context.getBuildFields();
  }
}
```

**Comment by RealyUniqueName** [comment]
Still not reproducible. What's your Haxe version and which compilation arguments do you use?

**Comment by lublak** [comment]
Arguments:
```hxml
-cp src
-D analyzer-optimize
--dce full
-main Main
--interp
```
I use vscode for the development
And i use haxe 4.2.3

**Comment by RealyUniqueName** [comment]
I managed to to reproduce the issue by building multiple times through compilation server and commenting/uncommenting `new Test();` between compilations. Without `Test.hx` the issue is not reproducible no matter how I change `Main.hx` between compilations.
I guess the root of the issue is that your code defines a module, which already exists (`Test.hx`).
@Simn shouldn't `Context.defineModule(modulePath, ...)` fail if `modulePath` already exists?

**Comment by Simn** [comment]
> shouldn't Context.defineModule(modulePath, ...) fail if modulePath already exists?

Yeah I'd expect the infamous "module redefined" error in that case.

**Comment by lublak** [comment]
But than: How to inject in the same module?
I use it to inject in an existing module and it works fine with classes.

**Comment by RealyUniqueName** [comment]
https://api.haxe.org/haxe/macro/Context.html#defineType

**Comment by Simn** [comment]
Might be a test case for the future.

**Comment by kLabz** [comment]
> > shouldn't Context.defineModule(modulePath, ...) fail if modulePath already exists?
> 
> Yeah I'd expect the infamous "module redefined" error in that case.

That seems to contradict with https://github.com/HaxeFoundation/haxe/commit/e75dfdacf59a85e220d5c6f9e19a77a67babd4f0 :thinking: 

I can trigger the error with two cases:
* if Test has module level fields it will break at first compilation
* if not, it will break starting second compilation

I implemented proper error reporting when module already exists, but linked commit implies we actually want to allow this :thinking:
Do you know which kind of use cases would actually be able to do that successfully? (will run CI with my changes to see if it catches one)

**Comment by Simn** [comment]
I don't remember why I made that change and it doesn't look like a good idea, especially in the context of the compilation cache.

**Comment by kLabz** [comment]
Ok, hopefully if there's a good reason for this to be allowed it's present in our tests.
Let's see how this goes ~ https://github.com/kLabz/haxe/actions/runs/5011170647

**Comment by kLabz** [comment]
So hmm it's used explicitly in #4160 and #6006 with
```haxe
Context.defineModule(Context.getLocalModule(), [...]);
```

**Comment by Simn** [comment]
Well, I guess it's kind of useful to define a bunch of types into the same module, but it also seems like such a compilation order hassle...

**Comment by kLabz** [comment]
It also seems broken in some cases where module level fields are involved.
I could look into fixing this instead, but it's not worth it if we'd rather disallow the whole thing.
